### PR TITLE
[Hub Generated] Review request for Microsoft.Insights to add version stable/2024-02-01

### PR DESCRIPTION
**Pull request by ToddKingMSFT**
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

> [!NOTE]
> As of January 2024 there is no PR assignee. This is expected. See https://aka.ms/azsdk/pr-arm-review.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

diagram

<details>
<summary> Click here to see the details of Step 1 </summary>

### Breaking changes review (Diagram Step 1)

If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
- The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
- You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

</details>

<details> 
<summary> Click here to see the details of Step 2 </summary>

### ARM API changes review (Diagram Step 2)

- For help with Step 2 / "ARM review", see https://aka.ms/azsdk/pr-arm-review.

</details>

<details>
<summary> Click here to see the diagram footnotes</summary>

### Diagram footnotes

[1] See [ARM review queue] (for **PR merge** queues, see [2]).  
[2] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])  
The ARM reviewer on-call engineer visits the merge queue twice a day, so the approximate ETA for merges is 12 - 24 hours.

</details>

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been generated using [OpenAPI Hub](https://aka.ms/openapihub)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with PR workflow diagram Step 2 (ARM review), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://aka.ms/azsdk/pr-arm-review#about-the-arm-review-queue
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

### CODE REVIEW

**Comment by ToddKingMSFT** [comment]
> Hi @ToddKingMSFT! For review efficiency consideration, when creating a new API version, it is required to place API specs of the base version in the first commit, and push new version updates into successive commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

This was created using the OpenAPIHub

**Comment by maririos** [comment]
> > Hi @ToddKingMSFT! For review efficiency consideration, when creating a new API version, it is required to place API specs of the base version in the first commit, and push new version updates into successive commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).
> 
> This was created using the OpenAPIHub

The comment was triggered because the name of the PR was changed so label `CI-MissingBaseCommit` failed.
@konrad-jamrozik who owns that logic?

**Comment by ToddKingMSFT** [comment]
> > > Hi @ToddKingMSFT! For review efficiency consideration, when creating a new API version, it is required to place API specs of the base version in the first commit, and push new version updates into successive commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).
> > 
> > 
> > This was created using the OpenAPIHub
> 
> The comment was triggered because the name of the PR was changed so label `CI-MissingBaseCommit` failed. @konrad-jamrozik who owns that logic?

@maririos / @konrad-jamrozik I don't think I customized the branch name, as I recall I accepted the name one hub suggested unchanged.

**Comment by maririos** [comment]
> @maririos Mariana Rios Flores FTE / @konrad-jamrozik Konrad Jamrozik FTE I don't think I customized the branch name, as I recall I accepted the name one hub suggested unchanged.

@ToddKingMSFT I meant the tittle of the Pull request. Usually it has a prefix that the tooling is looking for. I am not saing this is right and people shouldn't modify the name of the PR, just highlighting why you got the failure.

**Comment by konrad-jamrozik** [comment]
@maririos

> The comment was triggered because the name of the PR was changed so label CI-MissingBaseCommit failed.
@konrad-jamrozik who owns that logic?

I do :)

**Comment by ToddKingMSFT** [comment]
Generate ApiView

File: specification/monitor/resource-manager/Microsoft.Insights/stable/2024-02-01/metricNamespaces_API.json [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        "type": {
+          "type": "string",
+          "description": "The type of the namespace."
+        },
+        "name": {
+          "type": "string",
+          "description": "The escaped name of the namespace."
+        },
+        "classification": {
+          "$ref": "#/definitions/NamespaceClassification",
```
Comment (rkmanda) [inline specification/monitor/resource-manager/Microsoft.Insights/stable/2024-02-01/metricNamespaces_API.json:118]: classification is not a top level property in ARM envelopes. pl move it inside the properties bag

Comment (rkmanda) [reply specification/monitor/resource-manager/Microsoft.Insights/stable/2024-02-01/metricNamespaces_API.json:118]: or did u mean for this to be the "kind" property?

Comment (ToddKingMSFT) [reply specification/monitor/resource-manager/Microsoft.Insights/stable/2024-02-01/metricNamespaces_API.json:118]: I can't change this, that would be a breaking change. I realize the diff tool is broken but the only change between this and the 2023-10-01 version is the version. This is just a version rev with no other changes.

**Comment by ToddKingMSFT** [comment]
I just wanted to add some more context on this, the whole point of this PR is our service has had the metricNamespaces in preview for 6 years (2017-12-01-preview) so it has become GA and needs to be exposed in a stable API version. We originally wanted to add it to our latest 2023-10-01 API version but discovered adding a new API to an existing API is a breaking change so we're introducing a new API just so we can bring all 3 metrics APIs (metrics, metricDefinitions, metricNamespaces) into a single API version.

The only changes in this PR are the version.
metrics and metricDefinitions are identical to the 2023-10-01 versions, and metricNamespaces is identical to the 2017-12-01-preview version.

Since this is just an API only version change we were hoping this would be a speedy review but 3 weeks later we are disappointed to see this still has not be approved.

**Comment by kazrael2119** [comment]
add @JeffreyRichter as reviewer because the default tag contains multi api versions

**Comment by ToddKingMSFT** [comment]
> Todd, what is the plan for getting all the sub-services to share the name api-version and not have preview intermingled with GA versions?

At this point I'm waiting on some kind of policy or documentation I can point to that clearly indicates this is wrong so I can raise this with the rest of my org.

**Comment by konrad-jamrozik** [comment]
> > Todd, what is the plan for getting all the sub-services to share the name api-version and not have preview intermingled with GA versions?
> 
> At this point I'm waiting on some kind of policy or documentation I can point to that clearly indicates this is wrong so I can raise this with the rest of my org.

@ToddKingMSFT

https://eng.ms/docs/products/azure-developer-experience/design/api-specs/api-versioning
https://aka.ms/AzBreakingChangesPolicy
https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes?anchor=best-practice/guidelines-on-common-scenarios
https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#api-versioning

**Comment by ToddKingMSFT** [comment]
> > > Todd, what is the plan for getting all the sub-services to share the name api-version and not have preview intermingled with GA versions?
> > 
> > 
> > At this point I'm waiting on some kind of policy or documentation I can point to that clearly indicates this is wrong so I can raise this with the rest of my org.
> 
> @ToddKingMSFT
> 
> * https://eng.ms/docs/products/azure-developer-experience/design/api-specs/api-versioning
> * https://aka.ms/AzBreakingChangesPolicy
> * https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes?anchor=best-practice/guidelines-on-common-scenarios
> * https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#api-versioning

We have a separate thread with Jeffrey and the Microsoft.Insights team on this. It's going to be a multi semester effort to resolve this. It's not something that can be done in a single PR.

**Comment by TimLovellSmith** [comment]
@JeffreyRichter Are you fine with merging this and following up on the versioning or other open issues later?

**Comment by ToddKingMSFT** [comment]
> @JeffreyRichter Are you fine with merging this and following up on the versioning or other open issues later?
@TimLovellSmith 
I just want to be clear, this change is making ZERO API changes, the only diff between these APIs and their previous versions is the api-version. The reason for this change is we ran afoul of a policy where we shouldn't release Track2 SDKs where multiple different API versions are used for different APIs so the purpose of this change is to get all the metrics APIs aligned on the same API version. 
This is currently blocking us from GAing the SDKs supporting a new metrics batch API many of our fortune 500 customers want. It has taken over a month to get something as simple as just incrementing an API version change in.
I understand the desire to get ALL of the APIs in the readme.md for the overall resource manager sdk aligned but this is a massive undertaking for Microsoft.Insights that we have a separate email thread going on and is going to take a lot of planning and will probably be a multi semester effort. 
In the meantime we can't abruptly block 100% of all Microsoft.Insights swagger changes until 100% of the api versions are aligned.
IMHO this change helps iteratively get us towards that goal by getting all of the metrics API versions aligned. We're going to have to iteratively fix this with changes like this vs doing it all in one fell swoop.

**Comment by JeffreyRichter** [comment]
Yes, I'm fine with approving this as things are no worse with everything staying as is and we can kick the can down the road a bit more before establishing a long-term fix.

**Review by TimLovellSmith — APPROVED** [review]

**Pull request merged**